### PR TITLE
Added bar overlap property for 2d bar charts

### PR DESCRIPTION
**Pull request by mindline-analytics**

### CODE REVIEW

File: src/PhpPresentation/Shape/Chart/Type/AbstractTypeBar.php [hunk 1]
```diff
@@ -127,6 +140,32 @@ public function setGapWidthPercent($gapWidthPercent)
         return $this;
     }
 
+    /**
+     * @return int
+     */
+    public function getOverlapWidthPercent()
```
Comment (Progi1984) [inline src/PhpPresentation/Shape/Chart/Type/AbstractTypeBar.php:146]:
```suggestion
    public function getOverlapWidthPercent(): int
```

File: src/PhpPresentation/Shape/Chart/Type/AbstractTypeBar.php [hunk 2]
```diff
@@ -127,6 +140,32 @@ public function setGapWidthPercent($gapWidthPercent)
...
+    {
+        return $this->overlapWidthPercent;
+    }
+
+    /**
+     * @param int $overlapWidthPercent
+     *
+     * @return $this
```
Comment (Progi1984) [inline src/PhpPresentation/Shape/Chart/Type/AbstractTypeBar.php:154]:
```suggestion
     * @return self
```

File: src/PhpPresentation/Shape/Chart/Type/AbstractTypeBar.php [hunk 3]
```diff
@@ -127,6 +140,32 @@ public function setGapWidthPercent($gapWidthPercent)
...
+     */
+    public function setOverlapWidthPercent($overlapWidthPercent)
```
Comment (Progi1984) [inline src/PhpPresentation/Shape/Chart/Type/AbstractTypeBar.php:156]:
```suggestion
    public function setOverlapWidthPercent(int $overlapWidthPercent): self
```

File: tests/PhpPresentation/Tests/Shape/Chart/Type/BarTest.php [hunk 4]
```diff
@@ -90,6 +90,32 @@ public function testGapWidthPercent(): void
...
+        $object->setBarGrouping(Bar::GROUPING_PERCENTSTACKED);
+        $this->assertEquals(100, $object->getOverlapWidthPercent());
+    }
+
+    public function testOverlapWidthPercent(): void
+    {
+        $value = mt_rand(-100, 100);
+        $object = new Bar();
+        $this->assertEquals(0, $object->getOverlapWidthPercent());
+        $this->assertInstanceOf('PhpOffice\\PhpPresentation\\Shape\\Chart\\Type\\Bar', $object->setOverlapWidthPercent($value));
```
Comment (Progi1984) [inline tests/PhpPresentation/Tests/Shape/Chart/Type/BarTest.php:111]:
```suggestion
        $this->assertInstanceOf(Bar::class, $object->setOverlapWidthPercent($value));
```

File: tests/PhpPresentation/Tests/Shape/Chart/Type/BarTest.php [hunk 5]
```diff
@@ -90,6 +90,32 @@ public function testGapWidthPercent(): void
...
+        $this->assertEquals($value, $object->getOverlapWidthPercent());
+        $this->assertInstanceOf('PhpOffice\\PhpPresentation\\Shape\\Chart\\Type\\Bar', $object->setOverlapWidthPercent(101));
```
Comment (Progi1984) [inline tests/PhpPresentation/Tests/Shape/Chart/Type/BarTest.php:113]:
```suggestion
        $this->assertInstanceOf(Bar::class, $object->setOverlapWidthPercent(101));
```

File: tests/PhpPresentation/Tests/Shape/Chart/Type/BarTest.php [hunk 6]
```diff
@@ -90,6 +90,32 @@ public function testGapWidthPercent(): void
...
+        $this->assertEquals(100, $object->getOverlapWidthPercent());
+        $this->assertInstanceOf('PhpOffice\\PhpPresentation\\Shape\\Chart\\Type\\Bar', $object->setOverlapWidthPercent(-101));
```
Comment (Progi1984) [inline tests/PhpPresentation/Tests/Shape/Chart/Type/BarTest.php:115]:
```suggestion
        $this->assertInstanceOf(Bar::class, $object->setOverlapWidthPercent(-101));
```

File: tests/PhpPresentation/Tests/Writer/PowerPoint2007/PptChartsTest.php [hunk 7]
```diff
@@ -603,6 +605,8 @@ public function testTypeBar(): void
         $this->assertZipXmlElementEquals('ppt/charts/' . $oShape->getIndexedFilename(), $element, $oSeries->getTitle());
         $element = '/c:chartSpace/c:chart/c:plotArea/c:barChart/c:gapWidth';
         $this->assertZipXmlAttributeEquals('ppt/charts/' . $oShape->getIndexedFilename(), $element, 'val', $valueGapWidthPercent);
+        $element = '/c:chartSpace/c:chart/c:plotArea/c:barChart/c:overlap';
+        $this->assertZipXmlAttributeEquals('ppt/charts/' . $oShape->getIndexedFilename(), $element, 'val', $valueOverlapWidthPercent);
```
Comment (Progi1984) [inline tests/PhpPresentation/Tests/Writer/PowerPoint2007/PptChartsTest.php:609]:
```suggestion
        $this->assertZipXmlAttributeEquals('ppt/charts/' . $oShape->getIndexedFilename(), $element, 'val', $oBar->getOverlapWidthPercent());
```

File: tests/PhpPresentation/Tests/Writer/PowerPoint2007/PptChartsTest.php [hunk 8]
```diff
@@ -582,6 +582,8 @@ public function testTypeBar(): void
         $oShape->setResizeProportional(false)->setHeight(550)->setWidth(700)->setOffsetX(120)->setOffsetY(80);
         $oBar = new Bar();
         $oBar->setGapWidthPercent($valueGapWidthPercent);
+        $valueOverlapWidthPercent = -10;
+        $oBar->setOverlapWidthPercent($valueOverlapWidthPercent);
```
Comment (Progi1984) [inline tests/PhpPresentation/Tests/Writer/PowerPoint2007/PptChartsTest.php:586]:
```suggestion
        $oBar->setOverlapWidthPercent(-10);
```

File: src/PhpPresentation/Shape/Chart/Type/AbstractTypeBar.php [hunk 9]
```diff
@@ -127,6 +140,32 @@ public function setGapWidthPercent($gapWidthPercent)
...
+    {
+        if ($overlapWidthPercent < -100) {
+            $overlapWidthPercent = -100;
+        }
+        if ($overlapWidthPercent > 100) {
+            $overlapWidthPercent = 100;
+        }
+        $this->overlapWidthPercent = $overlapWidthPercent;
```
Comment (Progi1984) [inline src/PhpPresentation/Shape/Chart/Type/AbstractTypeBar.php:164]:
```suggestion
        if ($value < -100) {
            $value = -100;
        }
        if ($value > 100) {
            $value = 100;
        }
        $this->overlapWidthPercent = $value;
```

File: docs/usage/shapes/chart.md [hunk 10]
```diff
@@ -331,6 +331,20 @@ $barChart = new Bar();
 $barChart->setGapWidthPercent(250);
 ```
 
+#### Overlap
+
+You can define the bar overlap within bars or columns clusters. It is defined in percent.
+The default value is 100% for stacked and 0% for grouped bar charts. The value must be defined between -100 and 100.
```
Comment (Progi1984) [inline docs/usage/shapes/chart.md:337]:
```suggestion
The default value is `100%` for stacked and `0%` for grouped bar charts. The value must be defined between -100 and 100.
```

**Review by Progi1984 — COMMENTED** [review]
It's a great job. Could you replace `overlapWidthPercent` by `overlap` ?

Don't forget to add your change to `1.0.0.md` changelog.

Could you give a screenshot of the old/new display with this method ?

**Comment by mindline-analytics** [comment]
Hi, thanks for the quick review!

By replacing overlapWidthPercent through overlap you mean the method names?

By screenshot you mean of a bar chart with/ without the overlap? And where should these be added?

I'll add the change to the changelog.

**Comment by Progi1984** [comment]
@mindline-analytics Thank you for your contribution.

Affirmative for replacing overlapWidthPercent through overlap you mean the method and attribute names.

Do you think you can add the screenshot to the documentation ?

**Comment by mindline-analytics** [comment]
Marriages have probably already failed because of methods and property naming.

Is there a plan to change the method/property name for the gap width? Since this is called `gapWidthPercent`, I would otherwise - for the sake of consistency - suggest to leave the naming as is. But it's your decision, if desired, I'll be happy to change it of course. :)

File: docs/usage/shapes/chart.md [hunk 11]
```diff
@@ -14,7 +14,7 @@ $chartShape = $slide->createChartShape();
 
 You can define how blank values are displayed with the method `setDisplayBlankAs`.
 
-![Slideshow type](/images/libreoffice_chart_displayblankas.png)
+![Slideshow type](../../images/libreoffice_chart_displayblankas.png)
```
Comment (Progi1984) [inline docs/usage/shapes/chart.md:17]:
```suggestion
![Slideshow type](/images/libreoffice_chart_displayblankas.png)
```

File: docs/usage/shapes/chart.md [hunk 12]
```diff
@@ -321,16 +321,37 @@ TODO
...
+$barChart->setBarGrouping(Bar::GROUPING_CLUSTERED);
+
+// a positive value will result in an overlapping
+$barChart->setOverlapWidthPercent(25);
+
+// a negative value will result in a gap
+$barChart->setOverlapWidthPercent(-25);
+```
+
+![Bar Overlap](../../images/chart_bar_overlap.png)
```
Comment (Progi1984) [inline docs/usage/shapes/chart.md:353]:
```suggestion
![Bar Overlap](/images/chart_bar_overlap.png)
```

File: docs/usage/shapes/chart.md [hunk 13]
```diff
@@ -349,18 +370,11 @@ $barChart->setBarGrouping(Bar::GROUPING_STACKED);
...
-
-- Bar::GROUPING_PERCENTSTACKED
-
-![Bar::GROUPING_PERCENTSTACKED](/images/chart_columnpercent_52x60.png)
-
+|                                                                             | Type         | Constant                     |
+| --------------------------------------------------------------------------- | ------------ | ---------------------------- |
+| ![Bar::GROUPING_CLUSTERED](../../images/chart_columns_52x60.png)            | Grouped Bars | Bar::GROUPING_CLUSTERED      |
+| ![Bar::GROUPING_STACKED](../../images/chart_columnstack_52x60.png)          | Stacked Bars | Bar::GROUPING_STACKED        |
+| ![Bar::GROUPING_PERCENTSTACKED](../../images/chart_columnpercent_52x60.png) | Grouped Bars | Bar::GROUPING_PERCENTSTACKED |
```
Comment (Progi1984) [inline docs/usage/shapes/chart.md:377]:
```suggestion
| ![Bar::GROUPING_CLUSTERED](/images/chart_columns_52x60.png)            | Grouped Bars | Bar::GROUPING_CLUSTERED      |
| ![Bar::GROUPING_STACKED](/images/chart_columnstack_52x60.png)          | Stacked Bars | Bar::GROUPING_STACKED        |
| ![Bar::GROUPING_PERCENTSTACKED](/images/chart_columnpercent_52x60.png) | Grouped Bars | Bar::GROUPING_PERCENTSTACKED |
```

**Comment by Progi1984** [comment]
@mindline-analytics OK for keeping naming. Last feedback on paths for images and I merge it.

**Comment by mindline-analytics** [comment]
@Progi1984: Why did you revert the image paths? Is this related to the mkdocs process? I'm not familiar with it, but currently the image sources are broken in [github](https://github.com/PHPOffice/PHPPresentation/blob/develop/docs/usage/shapes/chart.md) und [github.io](https://phpoffice.github.io/PHPPresentation/usage/shapes/chart.html).

**Review by Progi1984 — APPROVED** [review]

**Comment by Progi1984** [comment]
@mindline-analytics You're right. Thank you for your contribution.

**Pull request merged**